### DISCUSSION TITLE
Automatically choose 1st option in dependent Select component

### DISCUSSION THREAD

**Discussion by wickham-c** [filamentphp/filament#12074] [Help] [answered]
### Package

Form builder

### Package Version

v3.2.52

### How can we help you?

I've created a Filament form in which I have dependent Select components. 

I choose a "client" (i.e. a company name) from the 1st Select component (`client_id`) and this successfully populates the available options in the 2nd Select component (`contact_id`) by utilising the `modifyQueryUsing` capability in the relationship, to pass the chosen `client_id` into the query.

What I would like to do is automatically choose the 1st "contact" (i.e. employee of the client) option immediately after a "client" has been chosen. I have attempted to do this using `afterStateUpdated`. The `$set` function call does appear to successfully set `contact_id` to the desired `id` number (I've checked by temporarily dumping a call to `$get('contact_id')` on the line below the `$set` call), but the 2nd Select component doesn't visually show that this value has been chosen. Therefore, when attempting to submit the form, the 2nd Select component is highlighted as being "required".

Is there a straightforward way of automatically choosing the **1st option** in the 2nd Select component, once an option from the 1st Select component has been chosen? 

```
            Select::make('client_id')
                ->relationship('client', 'name')
                ->live()
                ->afterStateUpdated(function (Set $set, $state) {
                    $set('contact_id', Contact::firstWhere('client_id', $state)->id);
                })
                ->required()
                ->searchable()
                ->preload(),
            
            Select::make('contact_id')
                ->relationship('contact', 'full_name', modifyQueryUsing: function (Builder $query, Get $get) {
                    return $query->where('client_id', $get('client_id'));
                })
                ->hint(str('Select a **Client** to populate this drop-down list')->inlineMarkdown()->toHtmlString())
                ->required(),
```
Any help would be gratefully received.

**Comment by micraux** [comment]
Did you try [setting a default value](https://filamentphp.com/docs/3.x/forms/fields/getting-started#setting-a-default-value)?

**Reply by wickham-c** [reply]
I did, but...

1. I wasn't sure what to enter as the default value.
2. I believe that the default is set as the form loads. As the 2nd Select component (`contact_id`) isn't populated until the 1st Select component's (`client_id`) option has been chosen, the 2nd Select component doesn't have any values to default to when the form is loaded.

**Reply by leandrocfe** [reply, marked as answer]
```php
Select::make('client_id')
    ->relationship('client', 'name')
    ->live()
    ->afterStateUpdated(function (Select $component) {

        $select = $component->getContainer()->getComponent('contact');
        $select->state(array_key_first($select->getOptions()));

    })
    ->required()
    ->searchable()
    ->preload(),

Select::make('contact_id')
    ->key('contact')
    ->extraInputAttributes(['wire:key' => Str::random(10)])
    ->relationship('contact', 'full_name', modifyQueryUsing: function (Builder $query, Get $get) {
        return $query->where('client_id', $get('client_id'));
    })
    ->hint(str('Select a **Client** to populate this drop-down list')->inlineMarkdown()->toHtmlString())
    ->required(),
```

**Reply by wickham-c** [reply]
Many thanks @leandrocfe - that's done the trick. Have a good weekend.

**Reply by irfanandratama** [reply]
I am trying using this to select the 1st options but the value won't save. Can you help me out?

```
                Forms\Components\Select::make('room_id')
                    ->relationship('room', 'name')
                    ->preload()
                    ->live()
                    ->required()
                    ->afterStateUpdated(function (Select $component) {
                        $select = $component->getContainer()->getComponent('petugas');
                        $select->state(array_key_first($select->getOptions()));
                    })
                    ->label('Ruangan')
                    ->placeholder('Pilih ruangan')
                    ->disabled()
                    ->dehydrated(),
                Forms\Components\Select::make('assign_to')
                    ->key('petugas')
                    ->extraInputAttributes(['wire:key' => Str::random(10)])
                    ->relationship(
                        name: 'assign',
                        titleAttribute: 'name', 
                        modifyQueryUsing: fn (Builder $query, Get $get) => $query->join('room_users', 'users.id', '=', 
                        'room_users.user_id')->join('rooms', 'rooms.id', '=', 
                        'room_users.room_id')->where('room_users.room_id', $get('room_id')),
                    )
                    ->live()
                    ->preload()
                    ->label(__('Petugas Kebersihan'))
                    ->placeholder('Pilih petugas kebersihan')
                    ->selectablePlaceholder(false),
```
